icu4c/source/common/mlbe.cpp

```diff
@@ -53,28 +48,27 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
         return 0;
     }
     int32_t numBreaks = 0;
-    UnicodeString index;
-    // The ML model groups six char to evaluate if the 4th char is a breakpoint.
-    // Like a sliding window, the elementList removes the first char and appends the new char from
-    // inString in each iteration so that its size always remains at six.
-    UChar32 elementList[6];
-
-    int32_t codeUts = initElementList(inString, elementList, status);
-    int32_t length = inString.countChar32();
+    int32_t codePtSize = inString.countChar32();
 
+    // The ML model groups six char to evaluate if the 4th char is a breakpoint.
+    // Like a sliding window, the indexList removes the first char's index and appends the
+    // new char's index from inString in each iteration so that its size always remains at six.
+    int32_t indexList[6];
+    int32_t codeUts = initIndexList(inString, indexList, status);
     // Add a break for the start.
     boundary.addElement(0, status);
     numBreaks++;
     if (U_FAILURE(status)) return 0;
 
-    for (int32_t i = 1; i < length && U_SUCCESS(status); i++) {
-        evaluateBreakpoint(elementList, i, numBreaks, boundary, status);
-        if (i + 1 >= inString.countChar32()) break;
-        // Remove the first element and append a new element
-        uprv_memmove(elementList, elementList + 1, 5 * sizeof(UChar32));
-        elementList[5] = inString.countChar32(0, codeUts) < length ? inString.char32At(codeUts) : INVALID;
-        if (elementList[5] != INVALID) {
-            codeUts += U16_LENGTH(elementList[5]);
+    for (int32_t i = 1; i < codePtSize && U_SUCCESS(status); i++) {
+        evaluateBreakpoint(inString, indexList, i, codeUts, numBreaks, boundary, status);
+        if (i + 1 >= codePtSize) break;
+        uprv_memmove(indexList, indexList + 1, 5 * sizeof(int32_t));
```

Comment:
so.... my understanding is you use a fixed size 6 array on the stack, and going through the whole list codePtSize of codeUnits . but you did uprv_memmove codePtSize times to shift 5 element.
Have you try to dynamic allocate codePtSize + 5 items (or maybe +6 or +7) for the indexList in the beginning and delete at the end but then you do not need such shift by calling uprv_memmove but rather use a different index for to decide where to start looking the next 6 items.
it will use more heap memory during the calls but given the input is usually not too big that should not be a big issue, right?

Reply:
Done